certora/specs/EnumerableSet.spec

```
@@ -0,0 +1,223 @@
+        }
+    }
+
+invariant consistencyKey(bytes32 key)
+    at_(to_uint256(_indexOf(key) - 1)) == key
+    {
+        preserved remove(bytes32 otherKey) {
+            requireInvariant consistencyKey(otherKey);
+            requireInvariant atUniqueness(
+                to_uint256(_indexOf(key) - 1),
+                to_uint256(_indexOf(otherKey) - 1)
+            );
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: state changes                                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule stateChange(env e, bytes32 key) {
+    require sanity();
+    requireInvariant consistencyKey(key);
+
+    uint256 lengthBefore   = length();
+    bool    containsBefore = contains(key);
+
+    method f; calldataarg args; f(e, args);
+
+    uint256 lengthAfter   = length();
+    bool    containsAfter = contains(key);
+
+    assert lengthBefore != lengthAfter => (
+        (f.selector == add(bytes32).selector    && lengthAfter == lengthBefore + 1) ||
+        (f.selector == remove(bytes32).selector && lengthAfter == lengthBefore - 1)
+    );
+
+    assert containsBefore != containsAfter => (
+        (f.selector == add(bytes32).selector    && containsAfter) ||
+        (f.selector == remove(bytes32).selector && containsBefore)
+    );
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: at only returns values for index that are in scope.                                                           │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule outOfBound(uint256 index) {
+    at_@withrevert(index);
+
+    assert lastReverted <=> length() <= index;
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: add key to EnumerableSet if not already contained                                                            │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule add(bytes32 key, bytes32 otherKey) {
+    require sanity();
+
+    uint256 lengthBefore        = length();
+    bool    containsBefore      = contains(key);
+    bool    containsOtherBefore = contains(otherKey);
+
+    bool added = add@withrevert(key);
+    bool success = !lastReverted;
+
+    // liveness & immediate effect
+    assert success && contains(key);
+
+    // return value: added iff not contained
+    assert added <=> !containsBefore;
+
+    // effect: length increass iff added
+    assert length() == lengthBefore + to_uint256(added ? 1 : 0);
```

Official [Certora's recommendation](https://docs.certora.com/en/latest/docs/cvl/mathops.html?highlight=casting#mathematical-operations) is to _use `mathint` where possible_, so:

```suggestion
    assert length() == lengthBefore + to_mathint(added ? 1 : 0);
```

Also, I think there's an implicit casting by just:

```suggestion
    assert length() == lengthBefore + added ? 1 : 0;
```

But I'm not sure if it works given the ternary

---

It work without `to_mathint` or `to_uint256` ... but we still need to put `()`